### ISSUE TITLE
Better handle responses sent to multiple request addresses

### ISSUE THREAD

**Issue by HelenWDTK** [mysociety/alaveteli#7879]
At the moment, all responses that are sent to more than one address are ending up in the holding pen, requiring manual intervention by site administrators to deliver. This has significantly increased the amount of email traffic related to holding pen (up 63% on the previous month). These messages represented over 70% of responses ending up in the holding pen in the past calendar month.

**Comment by garethrees** [comment]
What mechanism is being used to send a response to multiple addresses? Is it mostly in the To, Cc, Bcc, or no particular pattern there?

**Comment by garethrees** [comment]
Also, is there a common number of requests these responses get sent to (e.g. "normally 5 or fewer") or is it completely variable? ("sometimes 2, sometimes 20+")?

**Comment by HelenWDTK** [comment]
> What mechanism is being used to send a response to multiple addresses? Is it mostly in the To, Cc, Bcc, or no particular pattern there?

Mainly To: and CC:

> Also, is there a common number of requests these responses get sent to (e.g. "normally 5 or fewer") or is it completely variable? ("sometimes 2, sometimes 20+")?

Completely variable.

Also, Alaveteli treats these in two different ways that I've not yet spotted a reliable pattern in. Sometimes two or more copies of the email to be redelivered will end up in the holding pen, which can be individually delivered to the various requests they were meant for (so 2 request addresses, 2 copies, one for each). Other times (more commonly) the response appears only once in the holding pen, and needs sending to multiple places using comma separated request ids.

**Comment by mdeuk** [comment]
> What mechanism is being used to send a response to multiple addresses? Is it mostly in the To, Cc, Bcc, or no particular pattern there?

In the last few I've had, it was all `To`.

Worth noting, perhaps, that some of them are public bodies answering more than once [^1] - so there is a potential for a little confusion.

[^1]: e.g. Councils, and Policing bodies, with shared services may well address their response to more than one request simultaneously. Less confusing for the software, than for the human! 🤖

**Comment by mdeuk** [comment]
> Sometimes two or more copies of the email to be redelivered will end up in the holding pen, which can be individually delivered to the various requests they were meant for (so 2 request addresses, 2 copies, one for each). 

This one caught me out yesterday… incoming-2403264 appeared in the holding pen twice, but had both request IDs. Unfortunately, I ended up delivering it twice.

**Comment by HelenWDTK** [comment]
It appears that messages where multiple request addresses are put in the to: field end up in the holding pen more than once, but that messages where the other addresses are cc'd end up there just once.

**Comment by alexander-griffen** [comment]
n.b. This is when requests are exact, and isn't referring to requests matched using the Guess mechanism

**Comment by gbp** [comment]
> It appears that messages where multiple request addresses are put in the to: field end up in the holding pen more than once, but that messages where the other addresses are cc'd end up there just once.

Having tested with multiple email addresses in the `To`, `CC` and `BCC` headers. It seems this might not always be the case. With both multiple `To` & `CC` addresses we saw multiple messages in holding pens, with `BCC` these correctly got delivered to the requests - which is as expected.

Each pair of duplicate messages share the same `Message-ID` value which we could use to determine if the message should be delivered or has already been delivered to the request.

So the pseudo code we need is basically: For each request email address detected in the headers, guess which request it belongs to, filter requests by if the guess meets our threshold. Loop over each requests one by one, look at incoming messages `Message-ID` headers and if the current email we're processing has a new/different `Message-ID` then deliver the email to the request.

It might be wise to start caching `Message-ID` on the incoming messages instances in the future to optimise delivery. This can be done in a separate PR later.

**Comment by gbp** [comment]
> n.b. This is when requests are exact, and isn't referring to requests matched using the Guess mechanism

I'm actually think our "exact" matching isn't needed an more as wouldn't a Guess with ID score of 1 and ID hash of 1 also be an "exact" match. I might refactor this out in a future PR, for now I would try implement the fix for this to take guesses into account too.
